### ISSUE TITLE
[RFC][Feature][Autoscaler][Core]Graceful draining of nodes while scale-down

### ISSUE THREAD

**Issue by asm582** [ray-project/ray#23522]
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

The current implementation of the drain is just `kill `as confirmed by @DmitriGekhtman on kuberay repo: https://github.com/ray-project/kuberay/pull/208#discussion_r835970677. 

We want to implement a graceful drain where we allow workers to finish executing all the submitted tasks. This can be done by ignoring such nodes in subsequent scheduling cycles then cleaning them up when idle in a scale-down scenario. 

### Use case

This feature will be useful in scale-down scenario for kuberay project

### Related issues

https://github.com/ray-project/kuberay/pull/208

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by DmitriGekhtman** [comment]
Thanks, @asm582 for opening the issue!
It's RFC for now, but I think it's very important.

Two interlocking questions
- What should Ray's implementation of "drain" be? (cc @rkooo567 @scv119)
- How should external scaling machinery (autoscaler, operators, etc.) use the drain api? (cc @wuisawesome @sriram-anyscale @yiranwang52 @me)

**Comment by asm582** [comment]
If this helps, I tried to work on such a policy here: https://github.com/ray-project/ray/pull/18761

**Comment by Jeffwan** [comment]
This would be super helpful. I am working on a proposal to proactively migrate workloads to healthy nodes to reduce workloads overall downtime. `drain` is required here to achieve this. Let me know if I can give any help

**Comment by rkooo567** [comment]
cc @scv119

**Comment by asm582** [comment]
@Jeffwan we tried adding a new scheduling state as referenced in here #18761, would be interesting to know your thoughts on it. our proposal for graceful scale-down is here #18576.

**Comment by DmitriGekhtman** [comment]
cc @wuisawesome

**Comment by wuisawesome** [comment]
upon slightly closer further inspection, this comment in the code details why we kill raylets:

```
      // NOTE(sang): Drain API is not supposed to kill the raylet, but we are doing
      // this until the proper "drain" behavior is implemented. Currently, before
      // raylet is killed, it sends a drain request to GCS. That said, this can
      // happen;
      // - GCS updates the drain state and kills a raylet gracefully.
      // - Raylet kills itself and send a drain request of itself to GCS.
      // - Drain request will become a no-op in GCS.
      // This behavior is redundant, but harmless. We'll keep this behavior until we
      // implement the right drain behavior for the simplicity. Check
      // https://github.com/ray-project/ray/pull/19350 for more details.
```

#19350 says `It's the implementation detail. Drain != kill raylet, but we are having this behavior for now because that's the least breaking change.` 

If I'm not mistaken, the autoscaler is the only user of this API and autoscaler folks are the ones who want to change this behavior. 

https://sourcegraph.com/search?q=context:%40Anyscale+DrainNode+-file:test+repo:ray&patternType=regexp&case=yes

@rkooo567 can you confirm? If so, it seems like this should be a pretty easy change.

**Comment by wuisawesome** [comment]
agreed though (1) seems a little difficult to do right now. Starting by eliminating (3) seems like a step in the right direction right? 

btw i assume when you say "migrate all the states from that node" you're referring to ownership info right?

**Comment by asm582** [comment]
@scv119  @wuisawesome one solution would be we ask users to confine actors with placement groups on the ray cluster, then we try to drain the node that has stateless tasks only, will that help simplify the design?

**Comment by wuisawesome** [comment]
> ask users to confine actors on certain nodes with placement groups on the ray cluster

I feel like we should separate mechanism from policy and drain node should be a mechanism. That may be a reasonable policy in some cases (in others, it may be infeasible if a workload is all actors or requires spreading, etc)

**Comment by rkooo567** [comment]
@wuisawesome you are right. @scv119 I feel like the right approach is going from (3) -> (2) -> (1) iteratively? (1) could be a pretty big project (moving objects & ownerships around could be more difficult than actors/tasks I believe)

**Comment by scv119** [comment]
hmm one thing i'm not sure is if we do 3 without fixing 1, what would the semantic of `drain`?

**Comment by wuisawesome** [comment]
> hmm one thing i'm not sure is if we do 3 without fixing 1, what would be the semantic of drain?

Good question. I think it would be equivalent to the `kubectl cordon` command (perhaps we should even rename it?). https://kubernetes.io/docs/concepts/architecture/nodes/

Do we have a roadmap for object migration and actor migration? Both seem like potentially difficult projects.

I think the "safe" downscaling routine in the meantime would be:

1. Pick a node
2. Remove it from scheduling
3. Wait for it to become idle
4. Remove it

Step 3 can have lots of complication. At the moment it just looks at the resources, but in the future, it could try to evict objects and actors, and also define "idle" to include 0-cpu actors and object ownership information.

**Comment by wuisawesome** [comment]
beep bop boop

**Comment by wuisawesome** [comment]
beep bop boop

**Comment by cszhu** [comment]
Hello! This P2 issue has seen no activity in at least 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
